File: machine/ts/machine/runtime/system_irq.ts
import { IRQ_NEWGAME, IRQ_REINIT } from '../bus/io';
import type { Runtime } from './runtime';

const SYSTEM_IRQ_MASK = (IRQ_REINIT | IRQ_NEWGAME) >>> 0;

export function raiseSystemIrq(runtime: Runtime, mask: number): void {
	const normalized = mask >>> 0;
	if (normalized === 0) {
		throw new Error('system IRQ mask must be non-zero.');
	}
	const unsupported = normalized & ~SYSTEM_IRQ_MASK;
	if (unsupported !== 0) {
		throw new Error(`unsupported system IRQ mask 0x${unsupported.toString(16)}.`);
	}
	runtime.machine.irqController.raise(normalized);
}

File: machine/cpp/src/machine/runtime/system_irq.cpp
#include "machine/runtime/system_irq.h"

#include "machine/bus/io.h"
#include "machine/runtime/runtime.h"

#include <stdexcept>

namespace bmsx {
namespace {

constexpr uint32_t SYSTEM_IRQ_MASK = IRQ_REINIT | IRQ_NEWGAME;

} // namespace

void raiseSystemIrq(Runtime& runtime, uint32_t mask) {
	if (mask == 0u) {
		throw BMSX_RUNTIME_ERROR("system IRQ mask must be non-zero.");
	}
	const uint32_t unsupported = mask & ~SYSTEM_IRQ_MASK;
	if (unsupported != 0u) {
		throw BMSX_RUNTIME_ERROR("unsupported system IRQ mask " + std::to_string(unsupported) + ".");
	}
	runtime.machine.irqController.raise(mask);
}

} // namespace bmsx
